server/api/tests/role.py

```diff
@@ -0,0 +1,802 @@
+            workspace=self.workspace,
+            member=self.member,
+            IS_OWNER=True,
+            MANAGE_WORKSPACE_MEMBERS=True,
+            MANAGE_WORKSPACE_ROLES=True,
+            MANAGE_SCHEDULES=True,
+            MANAGE_TIME_OFF=True
+        )
+
+        self.member2 = WorkspaceMember.objects.create(user=self.user2, workspace=self.workspace, added_by=self.user)
+        self.permissions2 = MemberPermissions.objects.create(
+            workspace=self.workspace,
+            member=self.member2,
+            IS_OWNER=False,
+            MANAGE_WORKSPACE_MEMBERS=False,
+            MANAGE_WORKSPACE_ROLES=False,
+            MANAGE_SCHEDULES=False,
+            MANAGE_TIME_OFF=False
+        )
+        self.client.force_authenticate(user=self.member.user)
+
+    def test_no_workspace(self):
+        data = {'name': 'test name1', 'pay_rate': 15.00}
+        response = self.client.put(self.url, data, format='json')
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_invalid_workspace(self):
+        data = {'workspace_id': 999, 'name': 'test name1', 'pay_rate': 15.00}
+        response = self.client.put(self.url, data, format='json')
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+
+    def test_create_workspace_role_valid(self):
+        data = {'workspace_id': self.workspace.id, 'name': 'test name1', 'pay_rate': 15.00}
+        response = self.client.put(self.url, data, format='json')
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        
+        role = WorkspaceRole.objects.get(workspace=self.workspace, name=data['name'])
+
+        self.assertEqual(role.workspace, self.workspace)
+        self.assertEqual(role.name, data['name'])
+        self.assertEqual(role.pay_rate, data['pay_rate'])
+
+    def test_create_workspace_role_no_name_or_payrate(self):
+        data = {'workspace_id': self.workspace.id}
+        response = self.client.put(self.url, data, format='json')
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        
+        role = WorkspaceRole.objects.get(workspace=self.workspace)
+
+        self.assertEqual(role.workspace, self.workspace)
+        self.assertEqual(role.name, 'Unnamed Role')
+        self.assertEqual(role.pay_rate, None)
+
+    def test_create_workspace_role_without_permissions(self):
+        data = {'workspace_id': self.workspace.id, 'name': 'test name1', 'pay_rate': 15.00}
+
+        self.client.force_authenticate(user=self.user2) # change to send request from user 2
+
+        response = self.client.put(self.url, data, format='json')
+
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+    def test_create_workspace_role_as_non_member(self):
+        data = {'workspace_id': self.workspace.id, 'name': 'test name1', 'pay_rate': 15.00}
+
+        self.client.force_authenticate(user=self.user3) # change to send request from user 2
```

Comment:
Does this need to be changed?

Reply:
creating workspace roles still needs workspace_id provided if thats what you meant

Reply:
wait you mean the auth; The auth is right, the comment is wrong